### DISCUSSION TITLE
Beta V1 Version of GeoJson Extraction

### DISCUSSION THREAD

**Discussion by 2320sharon** [SatelliteShorelines/CoastSeg#15] [Show and tell]
# Beta V1 Version of GeoJson Extraction
I made a small program using ipyleaflet to create an interactable map to download geojson for CoastSat's data retrieval.  This program's main function is to let the user draw a polygon and add its geojson to a list called "shapes_list", which would then be used to supply data to CoastSat's data retrieval.

- This program is a proof concept that I think could be augmented with Geemap and Leafmap for more advanced functionality.

- Note: I made a Jupyter notebook with all the code and the dataset that I cannot upload to this discussion post. Let me know if you would like it.

# The Code 
---

## Interactive Drawable Map with CoastSat's Geojson
- [CoastSat Github](https://github.com/kvos/CoastSat) 
- Website where geojson was collected [CoastSat Website](http://coastsat.wrl.unsw.edu.au/)

## Optional Code to Install json, ipyleaflet

```python
# pip install json
# pip install ipyleaflet
```

## Add Clustering to Points
1. Import all the necessary libraries to draw the map
2. Open CoastSat's geojson
3. Create a zoomed out Map
4. Load in the geojson with the GeoJSON function
    - choose the geojson to load in
    - style the geojson
    - add a hover style interaction
5. Make markers from the geojson data
6. Transform the markers into a cluster (for better loading + visibility ) and add them as a layer to the map called "CoastSat Shoreline Data"
7. Add layer selection control
8. Add a scale control
9. Display the map

![screenshot1_beta1](https://user-images.githubusercontent.com/61564689/146462506-8ddc6f05-2b9c-4efc-b94e-34ea070561da.png)

```python
import json
import random

from ipyleaflet import Map, GeoJSON, Icon, Marker, MarkerCluster, LayersControl, basemap_to_tiles ,basemaps

# Name of the geojson: metadata_Pacific_CoastSat.geojson
with open(r'C:\Users\Sharon\Desktop\metadata_Pacific_CoastSat.geojson', 'r') as f:
    data = json.load(f)

m = Map(center=(50.6252978589571, 0.34580993652344), zoom=1)

# Optional Code if you want to have a different color for each geojson feature
def random_color(feature):
    return {
        'color': 'black',
        'fillColor': random.choice(['red', 'yellow', 'green', 'orange']),
    }

geo_json = GeoJSON(
    data=data,
    style={
        'opacity': 1, 'dashArray': '8', 'fillOpacity': 0.1, 'weight': 1
    },
    hover_style={
        'color': 'white', 'dashArray': '0', 'fillOpacity': 0.2
    },
    style_callback=random_color     #This calls the random_color function to style each geojson feature
)

# This turns all of the marks into maps to make them easier to view
markers = ()
features = data['features']
# Map each locations latitde and longtitde to a marker then add it to the cluster as a tuple
for i in range(len(features)):
    location=(features[i]['geometry']['coordinates'][1],features[i]['geometry']['coordinates'][0])
    marker = Marker(location=location)
    markers = markers + (marker,)

m.add_layer(MarkerCluster(name = "CoastSat Shoreline Data",markers = markers,))

# Add another layer
dark_matter_layer = basemap_to_tiles(basemaps.Stamen.Terrain)
m.add_layer(dark_matter_layer)


# Add a way to control which layers are displayed
m.add_control(LayersControl())

# Add a scale to the map
from ipyleaflet import ScaleControl
m.add_control(ScaleControl(position='bottomleft'))

# Display the map
m

```

    Map(center=[50.6252978589571, 0.34580993652344], controls=(ZoomControl(options=['position', 'zoom_in_text', 'z…

## Add Polygon Draw Control
- Made it so the user can only draw polygons
- Each polygon the user draws is add to the shape_list
- IDEA: use the shape_list to provide data to CoastSat
- WARNING: if you run this cell twice in jupyter notebook it was a weird effect of adding an extra polygon button to the drawing control

![screenshot2_beta1](https://user-images.githubusercontent.com/61564689/146462534-bbc881e0-4634-47fb-8877-0edeb3b784cd.png)

```python
from ipyleaflet import DrawControl

shapes_list=[]

draw_control = DrawControl()
draw_control.polygon = {
    "shapeOptions": {
        "fillColor": "#6be5c3",
        "color": "#6be5c3",
        "fillOpacity": 0.4
    },
    "drawError": {
        "color": "#dd253b",
        "message": "Ops!"
    },
    "allowIntersection": False,
    "transform":True
}

# Disable polyline, circle, and rectangle 
draw_control.polyline = {}
draw_control.circlemarker = {}
draw_control.rectangle = {}

def handle_draw(target, action, geo_json):
    if draw_control.last_action == 'created':
        shapes_list.append( draw_control.last_draw['geometry'])
    print("\nshapes_list: ",shapes_list)




draw_control.on_draw(handle_draw)
m.add_control(draw_control)

m
```

    Map(center=[50.6252978589571, 0.34580993652344], controls=(ZoomControl(options=['position', 'zoom_in_text', 'z…

    
    shapes_list:  [{'type': 'Polygon', 'coordinates': [[[-188.659363, 51.495065], [-191.216476, 18.398144], [-127.289512, 31.129926], [-188.659363, 51.495065]]]}]
    

```python
print(shapes_list)
```

**Comment by ddbuscombe** [comment]
Excellent stuff. I will digest and test. I added you as a contributor to this repo, so you can upload your ipython notebook (thanks)
